src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Cohost/OpenDocumentGenerator.cs

```diff
@@ -0,0 +1,148 @@
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.Cohost;
+
+[Export(typeof(OpenDocumentGenerator)), Shared]
+[method: ImportingConstructor]
+internal sealed class OpenDocumentGenerator(
+    LanguageServerFeatureOptions languageServerFeatureOptions,
+    LSPDocumentManager documentManager,
+    CSharpVirtualDocumentAddListener csharpVirtualDocumentAddListener,
+    JoinableTaskContext joinableTaskContext,
+    IRazorLoggerFactory loggerFactory)
+{
+    private readonly LanguageServerFeatureOptions _languageServerFeatureOptions = languageServerFeatureOptions ?? throw new ArgumentNullException(nameof(languageServerFeatureOptions));
+    private readonly TrackingLSPDocumentManager _documentManager = documentManager as TrackingLSPDocumentManager ?? throw new ArgumentNullException(nameof(documentManager));
+    private readonly CSharpVirtualDocumentAddListener _csharpVirtualDocumentAddListener = csharpVirtualDocumentAddListener ?? throw new ArgumentNullException(nameof(csharpVirtualDocumentAddListener));
+    private readonly JoinableTaskContext _joinableTaskContext = joinableTaskContext;
+    private readonly ILogger _logger = loggerFactory.CreateLogger<OpenDocumentGenerator>();
+
+    public async Task DocumentOpenedOrChangedAsync(IDocumentSnapshot document, int version, CancellationToken cancellationToken)
+    {
+        // These flags exist to workaround things in VS Code, so bringing cohosting to VS Code without also fixing these flags, is very silly.
+        Debug.Assert(_languageServerFeatureOptions.IncludeProjectKeyInGeneratedFilePath);
+        Debug.Assert(!_languageServerFeatureOptions.UpdateBuffersForClosedDocuments);
+
+        // Actually do the generation
+        var generatedOutput = await document.GetGeneratedOutputAsync().ConfigureAwait(false);
+
+        // Now we have to update the LSP buffer etc.
+        // Fortunate this code will be removed in time
+        var hostDocumentUri = new Uri(document.FilePath);
+
+        _logger.LogDebug("[Cohost] Updating generated document buffers for {version} of {uri} in {projectKey}", version, hostDocumentUri, document.Project.Key);
+
+        if (_documentManager.TryGetDocument(hostDocumentUri, out var documentSnapshot))
+        {
+            // Html
+            var htmlVirtualDocumentSnapshot = TryGetHtmlSnapshot(documentSnapshot);
+
+            // CSharp
+            var csharpVirtualDocumentSnapshot = await TryGetCSharpSnapshotAsync(documentSnapshot, document.Project.Key, version, cancellationToken).ConfigureAwait(false);
+
+            // Buffer work has to be on the UI thread
+            await _joinableTaskContext.Factory.SwitchToMainThreadAsync(cancellationToken);
+
+            Debug.Assert(htmlVirtualDocumentSnapshot is not null && csharpVirtualDocumentSnapshot is not null ||
+                htmlVirtualDocumentSnapshot is null && csharpVirtualDocumentSnapshot is null, "Found a Html XOR a C# document. Expected both or neither.");
```

Comment:
>htmlVirtualDocumentSnapshot is null && csharpVirtualDocumentSnapshot is null [](http://example.com/codeflow?start=16&length=76)

just for my understanding - what kind of document would have neither buffer?

Reply:
None. If we have a host document, but no virtual documents, things are going very badly.

Reply:
Should we assert in that case as well then? I was just curious why we weren't asserting in that case and only in XOR case.

Reply:
Going to leave this with a Debug.Assert for now, as I expect this code will be deleted before we turn the feature flag on for any users.